### PR TITLE
crocodile/mwifiex

### PR DESCRIPTION
**Pull request by js731ca**
add support for the wifi module used with crocodile

with the upstream mwifiex driver made aware of the new sdio-id, and loading of the nxp/imx-firmware the modules wifi and bluetooth functions seem to work :-)

Note that the firmware does not come from linux-firmware.git but the nxp repository; which is AFAIK bad practice.. but (sadly?) already has precedence in: 562354ab9f0aa4fcd8f2184506dcb9c18a792182 - which also adds a new sdio id, but the referenced firmware binary is nowhere to be found :-\

### CODE REVIEW

File: drivers/net/wireless/marvell/mwifiex/sdio.c [hunk 1]
```diff
@@ -159,6 +159,59 @@ static const struct mwifiex_sdio_card_reg mwifiex_reg_sd8977 = {
...
+	.status_reg_0 = 0xE8,
+	.status_reg_1 = 0xE9,
+	.sdio_int_mask = 0xff,
+	.data_port_mask = 0xffffffff,
+	.io_port_0_reg = 0xE4,
+	.io_port_1_reg = 0xE5,
+	.io_port_2_reg = 0xE6,
+	.max_mp_regs = 196,
+	.rd_bitmap_l = 0x10,
+	.rd_bitmap_u = 0x11,
```
Comment (grembeter) [inline drivers/net/wireless/marvell/mwifiex/sdio.c:182]: Where all these values come from? Datasheet? Other mwifiex_reg structure?

Comment (js731ca) [reply drivers/net/wireless/marvell/mwifiex/sdio.c:182]: educated guess;
it's a copy of the datastructure for the "closest in name and function?" sd8997 - which seems to be also a sdio+uart wifi+bt combo chip 
and i didn't stumble over sth similar settings in the nxp driver... or at least nothing that seems different from how those two modules where handled there

File: arch/arm64/boot/dts/freescale/crocodile.dtsi [hunk 2]
```diff
@@ -349,6 +349,12 @@
 	non-removable;
 	mmc-pwrseq = <&usdhc1_pwrseq>;
 	status = "okay";
+
+	mwifiex: wifi@1 {
+		status = "okay";
```
Comment (bith3ad) [inline arch/arm64/boot/dts/freescale/crocodile.dtsi:354]: Can be dropped.

File: drivers/net/wireless/marvell/mwifiex/sdio.c (hunk 1 above)
Comment (bith3ad) [reply drivers/net/wireless/marvell/mwifiex/sdio.c:182]: Question is: are these registers, values, masks correct? Can you verify that? Or at least can add a pointer/link to  your commit message pointing to the location which you used as base?

File: drivers/net/wireless/marvell/mwifiex/sdio.c [hunk 3]
```diff
@@ -159,6 +159,59 @@ static const struct mwifiex_sdio_card_reg mwifiex_reg_sd8977 = {
...
+	.fw_dump_end = 0xf8,
+	.func1_dump_reg_start = 0x10,
+	.func1_dump_reg_end = 0x17,
+	.func1_scratch_reg = 0xe8,
+	.func1_spec_reg_num = 13,
+	.func1_spec_reg_table = {0x08, 0x58, 0x5C, 0x5D,
+				 0x60, 0x61, 0x62, 0x64,
+				 0x65, 0x66, 0x68, 0x69,
+				 0x6a},
+};
```
Comment (bith3ad) [inline drivers/net/wireless/marvell/mwifiex/sdio.c:213]: Same here.

**Review by bith3ad — APPROVED** [review]

File: drivers/net/wireless/marvell/mwifiex/sdio.c (hunk 3 above)
Comment (js731ca) [reply drivers/net/wireless/marvell/mwifiex/sdio.c:213]: please elaborate :-s

File: arch/arm64/boot/dts/freescale/crocodile.dtsi (hunk 2 above)
Comment (js731ca) [reply arch/arm64/boot/dts/freescale/crocodile.dtsi:354]: removed the 'status' with the next push

File: drivers/net/wireless/marvell/mwifiex/sdio.c (hunk 1 above)
Comment (js731ca) [reply drivers/net/wireless/marvell/mwifiex/sdio.c:182]: added a note in the commit message - but as said: educated guess = copy of the SD8997 structure; which seems to work just find :P

**Comment by js731ca** [comment]
there's still something minor missing: 
the wifi card only works (as in: is able to scan available networks, connect to one and then do a ping) 
*after* the module has been unloaded and loaded again once after boot:
```
root@crocodile:~# iw dev mlan0 scan                                                                                                                           
command failed: Device or resource busy (-16)                                                                                                                 
root@crocodile:~# modprobe -r mwifiex-sdio                                                                                                                    
[   51.217198] mwifiex_sdio mmc0:0001:1: info: shutdown mwifiex...                                                                                            
[   51.224305] mwifiex_sdio mmc0:0001:1: PREP_CMD: card is removed             
[   51.235140] mwifiex_sdio mmc0:0001:1: PREP_CMD: card is removed                                                                                            
[   51.294420] mwifiex_sdio mmc0:0001:1: PREP_CMD: card is removed    
[   51.306548] mwifiex_sdio mmc0:0001:1: PREP_CMD: card is removed                                                                                            
root@crocodile:~#                                                                                                                                             
root@crocodile:~# modprobe mwifiex-sdio                                                                                                                                                                                                                                                                                      
[   55.270405] mwifiex_sdio mmc0:0001:1: WLAN FW already running! Skip FW dnld
[   55.280230] mwifiex_sdio mmc0:0001:1: WLAN FW is active                                                                                                    
root@crocodile:~# [   55.313187] mwifiex_sdio mmc0:0001:1: Unknown api_id: 5                                                                                  
[   55.344609] mwifiex_sdio mmc0:0001:1: info: MWIFIEX VERSION: mwifiex 1.0 (16.92.21.p55)                                                                    
[   55.353637] mwifiex_sdio mmc0:0001:1: driver_version = mwifiex 1.0 (16.92.21.p55)                                                                          
root@crocodile:~# iw dev mlan0 scan                                                                                                                                                                                                                                                                                          
command failed: Device or resource busy (-16)                                                                                                                 
root@crocodile:~# iw dev mlan0 scan
<snip loong printout of available networks>          
```
the commands where sent successivly, less than a second apart - still investigating the "why", thoughts?

**Comment by js731ca** [comment]
found by thomas: ["[2/3] wifi: mwifiex: Support SD8978 chipset"](https://patchwork.kernel.org/project/linux-wireless/patch/536b4f17a72ca460ad1b07045757043fb0778988.1674827105.git.lukas@wunner.de/)

**Comment by js731ca** [comment]
the going-upstream patches have the same "resource busy" problem, but works "after a while"; with the exact same firmware binary referenced in the cover-letter "[0/3] mwifies material for next" ... i'll continue to investigate

**Comment by js731ca** [comment]
comparing to the out-of-tree kernel driver from NXP we've been using so far: the "resource busy" errors also pop up after a reboot; so it seems (!) to be a more general issue; and not only limited to mlan/moal (the NXP driver) or mwifiex/mwifiex-sdio (the upstream-going diver) 

suggestion: we wait for the patchset to pop up in upstream; -> closing this PR

**Pull request closed without merging**